Add 3D extent to providers (2nd stage)

benoitdm-oslandia opened the pull request:
This PR follows #54941. We have add 3D extent implementation for these providers:

* QgsOgrProvider
* QgsPostgresProvider
* QgsSpatiaLiteProvider
* QgsDelimitedTextProvider
* QgsPointCloudLayer 

These implementations could be discussed and the pointCloud one is a draft.

Funded by CEA/DAM @renardf

cc @ptitjano, @lbartoletti, @troopa81, @rouault, @uclaros

rouault commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1117,16 +1135,56 @@ unsigned char *QgsOgrProvider::getGeometryPointer( OGRFeatureH fet )
+      extent->MinY = std::min( extent->MinY, env.MinY );
+      extent->MaxX = std::max( extent->MaxX, env.MaxX );
+      extent->MaxY = std::max( extent->MaxY, env.MaxY );
+      extent->MinZ = std::min( extent->MinZ, env.MinZ );
+      extent->MaxZ = std::max( extent->MaxZ, env.MaxZ );
```
if there's a mix of 2D/3D objects in the layer, MinZ/MaxZ will be 0 on 2D objects

benoitdm-oslandia replied:
Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

May be to solve this, we can retrieve the 2D/3D state in each row?

benoitdm-oslandia replied:
I will try using `OGR_G_Is3D`

rouault replied:
> Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

I'm not sure of the reason besides being what is implemented... Note that NaN is also tricky to use with std::min() / std::max() (cf https://stackoverflow.com/questions/72296440/how-to-prevent-stdmin-and-max-to-return-nan-if-the-first-element-of-the-array)

> May be to solve this, we can retrieve the 2D/3D state in each row? I will try using OGR_G_Is3D

yes, that's what I had in mind

rouault commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1136,55 +1194,95 @@ QgsRectangle QgsOgrProvider::extent() const
+
+            mOgrLayer->GetExtent( mExtent.get(), true );
+            QByteArray geomColQuoted = quotedIdentifier( geomCol );
+            QByteArray sql = "SELECT MIN(ST_MinZ("
+                             + geomColQuoted
+                             + ")), MAX(ST_MaxZ("
+                             + geomColQuoted
+                             + ")) FROM "
+                             + quotedIdentifier( mOgrLayer->name() );
```
this will only work with PostGIS, GPKG and Spatialite datasets. You probably want to run the ExecuteSQL() between CPLPushErrorHandler(CPLQuietErrorHandler) / CPLPopErrorHandler() to avoid error messages on datasets that don't support that

rouault replied:
> The check `if ( l ) {...}` will be enough?

yes

rouault commented on tests/src/python/test_provider_ogr.py:
```diff
@@ -3497,6 +3498,41 @@ def testGeoJsonMapType(self):
+        self.assertFalse(os.path.exists(datasource_2d))
+
+        # 3D points
+        datasource_3d = os.path.join(self.basetestpath, 'testExtent3D.csv')
```
you'll probably also want to do a similar test on GeoPackage to the the "SELECT min(ST_MinZ(...) ..." code path

benoitdm-oslandia replied:
done!

nyalldawson commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -464,6 +464,24 @@ QgsOgrProvider::QgsOgrProvider( QString const &uri, const ProviderOptions &optio
+    bool is3D = true;
+    gdal::ogr_feature_unique_ptr f;
+
+    mOgrOrigLayer->ResetReading();
```
Why are we sniffing a feature here instead of just checking the wkb type of the layer?

rouault replied:
> Why are we sniffing a feature here instead of just checking the wkb type of the layer?

That was my initial reaction, but there are a number of data sources where the typing is loose. That said the strategy could be :
1. look at the wkb type of the layer: if it is 3D, then note it and stop
2. otherwise sniff a feature

benoitdm-oslandia replied:
fixed!

nyalldawson commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -247,6 +247,26 @@ QgsPostgresProvider::QgsPostgresProvider( QString const &uri, const ProviderOpti
+  // check if geometry is 3D or not
+  if ( !mGeometryColumn.isEmpty() && !mTableName.isEmpty() )
+  {
+    const QString selectQuery = QStringLiteral( R"SQL(
```
Again, why not just rely on the wkb type here?

benoitdm-oslandia replied:
fixed

benoitdm-oslandia replied:
We will reduce readability because the schemas/tables are not the same. This request is made once, the impact will be low.

nyalldawson replied:
I respectfully disagree -- see eg all the work done by https://github.com/qgis/QGIS/pulls?q=is%3Apr+is%3Aclosed+author%3Amorriganr focused on removing every possible extraneous query. That's driven by real world requirements.

benoitdm-oslandia replied:
Ok but I don't know how to properly do this merged query. Maybe @MorriganR could help us find a solution (please :grin:)?

nyalldawson replied:
Can we drop the postgres changes for now then, and implement in a specific follow up PR? I don't want to see speed regressions introduced here.

benoitdm-oslandia replied:
Yes it is the best solution for now

benoitdm-oslandia commented:
I just fix a regression in my spatialite. Tests should pass now!

benoitdm-oslandia commented:
@nyalldawson @rouault Tests are ok expect one with bad remote server access.

benoitdm-oslandia commented:
rebase from master...

uclaros commented:
Not a proper review, but you've left a couple of TODO comments there.
TBH, I find it quite hard to follow the review process when force-pushing...

rouault commented:
@benoitdm-oslandia Now that https://github.com/OSGeo/gdal/commit/8655f56595ca4c199069d6654ded11503d4f9802 has landed into GDAL master, it could be good that you have a GDAL >= 3.9.0 code path that uses OGR_L_GetExtent3D()

rouault commented:
> Does it work for all driver?

yes, there's a base implementation that will be the fallback for all drivers does what you implemented: iterating over all features. For shapefile, there's a specialized implementation that just gets the extent from the shapefile header.  GeoPackage and PostGIS specialized implementations are also planned.

nyalldawson commented:
@benoitdm-oslandia the test failure here ISN'T a false positive -- that test needs updating to match these changes

benoitdm-oslandia commented:
@nyalldawson my bad! Still works to do!

benoitdm-oslandia commented:
@rouault does your implementation handle the `mSubsetString` or the new `OGR_L_GetExtent3D` function will only replace the code from line 1230 to 1263 (main call to OGR_L_GetExtent)? will it handle the case from line 1211 to 1224 (ie. WITHOUT OLCFastGetExtent)?

rouault commented:
I'm CC'ing @elpaso who is implementing the functionality

> does your implementation handle the `mSubsetString` or the new `OGR_L_GetExtent3D` function will only replace the code from line 1230 to 1263 (main call to OGR_L_GetExtent)?

As far as I can see, that will be similar to GetExtent(). It is driver dependent whether they take into account or not filters. The generic (slow) implementation and as far as I can see the GeoPackage do, but the shapefile one does not. We could potentially make the shapefile implementation fallback to the generic (slow) implementation if there's a filter, and if taking it into account is desired.

> will it handle the case from line 1211 to 1224 (ie. WITHOUT OLCFastGetExtent)?

If the question is: will the OAPIF driver return a dummy extent when GetExtent3D() is called on it. The answer is: no, it won't, it will just go through the generic implementation.

benoitdm-oslandia commented:
@rouault I add a fix without checking the tests. Does it look ok to you?

rouault commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1222,72 +1222,42 @@ QgsBox3D QgsOgrProvider::extent3D() const
+        QgsDebugMsgLevel( QStringLiteral( "WITH OLCFastGetExtent" ), 3 );
+        QRecursiveMutex *mutex = nullptr;
+        OGRLayerH layer = mOgrLayer->getHandleAndMutex( mutex );
+        const char *geomCol = OGR_L_GetGeometryColumn( layer );
+        if ( geomCol != NULL && strlen( geomCol ) > 0 )
```
you don't need that test anymore as you don't write SQL. You could just test OGR_L_GetGeomType() != wkbNone

benoitdm-oslandia replied:
ok fixed

rouault commented on src/core/providers/ogr/qgsogrproviderutils.cpp:
```diff
@@ -3106,6 +3106,24 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,9,0)
+  return OGR_L_GetExtent3D( hLayer, psExtent3D, bForce );
+#else
+  OGREnvelope psExtent;
```
you should probably fallback to compute3DExtentSlowly() to get same results in QGIS whatever the GDAL version

benoitdm-oslandia replied:
done

rouault commented:
> I add a fix without checking the tests

What do you mean?

benoitdm-oslandia commented:
rebase from master, handle future gdal 3.9 extend3D function and fix tests

benoitdm-oslandia committed:
fix(QgsVectorLayer): extent2D/3D is computed according to real data elevation property

When the underlying data is 3D and 2D extent is first requested, the previous behaviour
was to retrieve and save the 2D extend. 3D extent data were lost.

benoitdm-oslandia committed:
feat(QgsOgrProvider): handle 3D extent/boundingbox

rework extent3D to handle 2D/3D, W/WO mSubsetString, W/WO OLCFastGetExtent

benoitdm-oslandia committed:
fix(QgsDelimitedTextProvider): improve 3D data handling

* Read from data if geometry has elevation and improve extent3D response.
* fix tests

benoitdm-oslandia committed:
feat(QgsPointCloudLayer): PARTIAL handle 3D extent/boundingbox

benoitdm-oslandia committed:
feat(QgsSpatiaLiteProvider): handle 3D extent/boundingbox

* Read from data if geometry has elevation and improve extent3D response.
* fix tests

benoitdm-oslandia committed:
QgsOgrProvider: use geom type (if any) to compute elevationProperties()->setContainsElevationData

benoitdm-oslandia committed:
QgsOgrProvider: protect 3D extent computation from unsupported sql

benoitdm-oslandia committed:
fix(QgsOgrProvider): add extent test for sqlite

benoitdm-oslandia committed:
fix(QgsOgrProvider): add extent test for gpkg

benoitdm-oslandia committed:
QgsOgrProvider: exclude 2D features from 3D extent

benoitdm-oslandia committed:
fix(test_provider_ogr): split extent test to extentCsv and extentShp

benoitdm-oslandia committed:
fix(TestQgsVectorDataProvider): handle 3D extent capacity of underlying provider

benoitdm-oslandia committed:
fix(ogrprovider): add support for GDAL (3.9.0) extent3D

* Move compute3DExtentSlowly to QgsOgrLayer
* Add new QgsOgrLayer::GetExtent3D function. For GDAL < 3.9, it returns a OGREnvelope3D based on OGR_L_GetExtent + sql request to retrieve MinZ/maxZ or fallback on computeExtent3DSlowly. If GDAL >= 3.9, it will use the OGR_L_GetExtent3D.
* QgsOgrProvider::extent3D is adapted accordingly

benoitdm-oslandia committed:
fix(test_provider_ogr.py): since the elevation property check in QgsOgrProvider, testHTTPRequestsOverrider needs a new request to check against.

Add comment to explain from what code comes the requests.

benoitdm-oslandia commented:
fix authors...

benoitdm-oslandia commented:
@rouault I have a question about gdal. 

Context: with `orginfo -al tests/testdata/points.shp` I obtain an extent with xmin=-118.8 but there is no point in this file with such coordinate. 

Question: does `ogrinfo` re-compute the extent each time or does it read it from the metadata of the shapefile? Is there a way to force the re-computation of the extent?

nyalldawson commented:
@benoitdm-oslandia regarding [695feac](https://github.com/qgis/QGIS/pull/55320/commits/695feacacdaac0398eecd78aab8a3a9b8be11b1a) -- what's triggering the new request? I don't think we should be calculating the 3d extent on layer load there, rather it should be deferred till it's actually required.

benoitdm-oslandia commented:
@nyalldawson can you give the right permission on GitHub to restart a failed job please?

benoitdm-oslandia commented:
> @benoitdm-oslandia regarding [695feac](https://github.com/qgis/QGIS/pull/55320/commits/695feacacdaac0398eecd78aab8a3a9b8be11b1a) -- what's triggering the new request? I don't think we should be calculating the 3d extent on layer load there, rather it should be deferred till it's actually required.

@nyalldawson, we are just reading the first feature to compute if the layer is 2D or 3D. We do not calculate the extent on layer load.

All lights are green and I am all done! Tell me if you have some point to discuss!

rouault commented on src/core/providers/ogr/qgsogrproviderutils.cpp:
```diff
@@ -3106,6 +3106,92 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
+{
+  QMutexLocker locker( &ds->mutex );
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,9,0)
+  return OGR_L_GetExtent3D( hLayer, psExtent3D, bForce );
```
```suggestion
  return OGR_L_GetExtent3D( hLayer, /* iGeomField = */ 0, psExtent3D, bForce );
```

rouault commented:
> Are you ok if we merge?

can you give it a try with GDAL master locally, and check the test_provider_ogr*.py tests pass  ?

uclaros commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1117,16 +1139,32 @@ unsigned char *QgsOgrProvider::getGeometryPointer( OGRFeatureH fet )
+  return extent3D().toRectangle();
+}
+
+// ================================== TODO: NOT CONST AT ALL!!!!
```
todo?

benoitdm-oslandia replied:
fixed! the extent member is mutable!

benoitdm-oslandia commented:
@rouault 

> can you give it a try with GDAL master locally, and check the test_provider_ogr*.py tests pass ?

The tests does not pass when the subset (attribute filter) is defined... Still checking why!

benoitdm-oslandia commented:
@rouault here is the path taken from qgis by asking `getExtent3D` on a 2D gpkg: 
1. `QgsOgrProvider::extent3D`
2. `QgsOgrLayer::GetExtent3D`
3. `OGR_L_GetExtent3D`
4. `OGRGeoPackageTableLayer::GetExtent3D` ==> as we are 2D we switch to 2D version
5. `OGRLayer::GetExtent`
6. `OGRGeoPackageTableLayer::GetExtent` ==> does not handle subset (aka attribute filter) ==> test failure

As we have said lately, `OGRGeoPackageTableLayer::GetExtent` will not handle the subset to avoid to break the existing behaviour. So I will add a check (qgis side) if the layer is 2D and if there is a subset defined, I will use the slow computed version. Can I suppose `QgsOgrLayer::GetExtent3D` implementations will handle the subset?

benoitdm-oslandia committed:
fix(OGRProvider): support for gdal 3.9 was buggy

As gdal OGRLayer::GetExtent implementation will not handle the subset to avoid to break
the existing behaviour. So we add a check (qgis side) if the layer is 2D and if there is a
subset defined, we will use the slow computed version. In other cases, we will delegate to
gdal with GetExtent3D.

rouault commented:
> As we have said lately, `OGRGeoPackageTableLayer::GetExtent` will not handle the subset to avoid to break the existing behaviour.

should now be fixed in GDAL master per https://github.com/OSGeo/gdal/commit/d26bb2861584d2588adc96478397417f4657df57

benoitdm-oslandia commented:
@rouault We pushed at the same time! I re-test with your new version.

benoitdm-oslandia commented:
Good for me!

benoitdm-oslandia committed:
feat(QgsPostgresProvider): handle 3D extent/boundingbox

* Read from data if geometry has elevation and improve extent3D response.
* Use geom type value (if any) or postgis geom type to compute elevationProperties()->setContainsElevationData
* Add tests

benoitdm-oslandia commented:
Add postgresql implementation

benoitdm-oslandia commented:
@nyalldawson @rouault everything is ok for you? Can we merge?

rouault commented:
OK for me (mostly reviewed the OGR part)

benoitdm-oslandia commented:
@nyalldawson as this PR has been approved, can you merge it please? Thanks!

lbartoletti commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -1082,14 +1088,24 @@ bool QgsPostgresProvider::loadFields()
+      if ( mDetectedGeomType == Qgis::WkbType::Unknown )
+      {
+        // rare case, elevation can not be computed via mDetectedGeomType, we parse the formattedFieldType and look for Z data
+        const thread_local QRegularExpression re( QRegularExpression::anchoredPattern( "geometry\\(\\w+ZM?(,\\d+)?\\)" ), QRegularExpression::CaseInsensitiveOption );
```
```suggestion
        const thread_local QRegularExpression re( QRegularExpression::anchoredPattern( "geometry\\(\\w+\s*Z?M?(,\\d+)?\\)" ), QRegularExpression::CaseInsensitiveOption );
```

Suggested changes:
- \s* : Sometimes WKT-like pattern is "POINT ZM" and not "POINTZM"
- Z?: Add a "?" here. Because, without "?", you only look for "Z" type. But, maybe it's normal?

ptitjano replied:
Applied. Thanks.

benoitdm-oslandia replied:
> Z?: Add a "?" here. Because, without "?", you only look for "Z" type. But, maybe it's normal?

Yes it was intentional!

lbartoletti commented on src/providers/postgres/raster/qgspostgresrastershareddata.cpp:
```diff
@@ -304,7 +304,7 @@ bool QgsPostgresRasterSharedData::fetchTilesIndex( const QgsGeometry &requestPol
   if ( !overallExtent.isNull() )
   {
-    mLoadedIndexBounds[ cacheKey ] = requestPolygon.combine( QgsGeometry::fromRect( overallExtent ) );
+    mLoadedIndexBounds[ cacheKey ] = requestPolygon.combine( QgsGeometry::fromWkt( overallExtent.asWktPolygon() ) );
```
Why use `fromWkt`? You can “lose” precision during the conversion double->str->double

ptitjano replied:
Thanks. This was a regression from a bad conflict resolution while rebasing the PR.

lbartoletti commented on tests/src/python/test_qgsdelimitedtextprovider.py:
```diff
@@ -878,6 +879,8 @@ def test_044_ZM(self):
         assert vl.isValid(), f"{basetestfile} is invalid"
         assert vl.wkbType() == QgsWkbTypes.PointZM, "wrong wkb type, should be PointZM"
         assert vl.getFeature(2).geometry().asWkt() == "PointZM (-71.12300000000000466 78.23000000000000398 1 2)", "wrong PointZM geometry"
+        self.assertEqual(vl.extent(), QgsRectangle(-71.12300000000000466, 66.32999999999999829, -65.31999999999999318, 78.29999999999999716))
```
```suggestion
        self.assertEqual(vl.extent(), QgsRectangle(-71.12300000000000466, 66.32999999999999829, -65.31999999999999318, 78.29999999999999716))
```

Theses tests will be flaky on different machines due to precision. Hope, you can use https://github.com/qgis/QGIS/pull/55648

ptitjano replied:
Indeed. I have switched to `self.assertAlmostEqual`  on each coordinate in the meantime. This is how the other providers test the extent.

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.h:
```diff
@@ -126,9 +126,9 @@ class QgsPostgresRasterProvider : public QgsRasterDataProvider
     //! Raster size x
-    qlonglong mWidth = 0;
+    long mWidth = 0;
     //! Raster size y
-    qlonglong mHeight = 0;
+    long mHeight = 0;
```
why this change?

benoitdm-oslandia replied:
linked to clang check

nyalldawson commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2212,7 +2227,11 @@ bool QgsPostgresProvider::parseDomainCheckConstraint( QStringList &enumValues, c
       if ( arrayPosition != -1 )
       {
         QString valueList = checkDefinition.mid( arrayPosition + 6, closingBracketPos );
+#if QT_VERSION < QT_VERSION_CHECK(5, 15, 0)
+        const QStringList commaSeparation = valueList.split( ',', QString::SkipEmptyParts );
+#else
```
this can be dropped, we require 5.15 now

benoitdm-oslandia replied:
indeed

nyalldawson commented on src/core/providers/ogr/qgsogrproviderutils.cpp:
```diff
@@ -3106,6 +3106,92 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
+  if ( err != OGRERR_NONE )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "Will computeExtent3DSlowly!" ), 3 );
+    err = computeExtent3DSlowly( psExtent3D );
```
Don't we still have a performance regression here? 

Ie in the situation:

- a driver which doesn't support fast 3d extent retrieval from OGR_L_GetExtent3D OR on older GDAL and the SELECT ST_MinZ... fails. 
- dataset has z
- the driver DOES support FAST retrieval of 2 extent
- ONLY the api to retrieve the 2d extent is called (eg to zoom to layer extent on loading the layer)

In that situation we're now forcing a complete table traversal when asking the provider for the 2d extent, because the 3d extent is calculated first. Whereas before we could quickly retrieve the 2d extent without table traversal...

benoitdm-oslandia replied:
hi @nyalldawson,

if the layer has z data and we have the 2D fast retrieval process for this driver, we still have to retrieve the minZ and maxZ data by using something like `SELECT ST_min(Z), ST_max(Z) FROM table`, which scan all rows. Also, by default for gdal 3.9, if the driver does not have an OGR_L_GetExtent3D implementation, gdal will scan all rows to compute the data (this is the same for the 2D extent).

Even if we use a second field to store the 2d extent when only the 2d extent is asked for a 3d layer, at some point, the 3D extent could be asked and we will have to scan all rows to compute it.

@rouault what do you think?

rouault replied:
> @rouault what do you think?

Will Extent3D will be called when adding a new vector layer ? If so, that might break some use cases. For example, currently you can add ``/vsizip//vsicurl/https://data.3dbag.nl/gpkg/v20231008/3dbag_nl.gpkg.zip`` (cf https://3dbag.nl/en/download . 19 GB as a ZIP) and use it reasonably with QGIS. But if Extent3D() is asked on it, then this will stall QGIS for hours.

nyalldawson replied:
> Will Extent3D will be called when adding a new vector layer 

Yes it will, because `QgsOgrProvider::extent()` just calls `return extent3D().toRectangle();`, and the layer extent is always required when adding a layer to a project (eg to zoom to the first layer added's extent in a new project)

nyalldawson replied:
Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations (ie not when adding layers or opening the layer properties windows).

rouault replied:
> Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations

I also concur with this

benoitdm-oslandia replied:
I have updated the PR to:
* use the old extent 2D code
* use 2 different field members to save the 2D extent and the 3D extent

The behaviour is now what you expected: we only load 3D extent when the 3D extent is asked.

Furthermore we have found a strange behaviour: 
a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.   
In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

This behaviour is not induced by this PR but just "revealed".

rouault replied:
> I have updated the PR to:
> 
>     * use the old extent 2D code
> 
>     * use 2 different field members to save the 2D extent and the 3D extent

LGTM !

> Furthermore we have found a strange behaviour:
> a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.
> In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

Ok, that's interesting. That lod12_2D layer doesn't set the Z=1 or 2 hint in the gpkg_geometry_columns, which is used by the driver to infer the dimensionality of the layer geometry type. I've fixed the writing side of the driver to set Z=2 (may contain 3D geoms) when one is encountered in https://github.com/OSGeo/gdal/pull/9147 . That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

benoitdm-oslandia replied:
> That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

indeed...

benoitdm-oslandia committed:
fix(QgsPostgresProvider): remove qt version check

As we are now, at least, using the Qt 5.15

benoitdm-oslandia committed:
Revert(QgsPostgresRasterProvider): clang fix for qlonglong

benoitdm-oslandia committed:
Fix(QgsOgrProvider): revert the old extent() code and add new field member to hold the 3D extent.

In order to benefit from the existing 2D optimization, we keep the 2D extent computation
as default. The 3D extent is only computed when the extent3D is called.

Also we will first believe in the metadata, even if the metadata is corrupted or misdefined.
We will not analyse the first geometry to compute if we have 3D or 2D data.

nyalldawson commented on src/core/vector/qgsvectorlayer.cpp:
```diff
@@ -1018,6 +1018,12 @@ QgsRectangle QgsVectorLayer::extent() const
   QGIS_PROTECT_QOBJECT_THREAD_ACCESS
 
+  // if data is 3D, redirect to 3D extend computation, and save it as 3D extent
+  if ( mDataProvider && mDataProvider->elevationProperties() && mDataProvider->elevationProperties()->containsElevationData() )
+  {
+    return extent3D().toRectangle();
+  }
```
This change here should be revert too right, otherwise we're again potentially calculating an expensive 3d extent when we only need the 2d extent.

benoitdm-oslandia replied:
my bad!

nyalldawson commented:
Nearly there :+1:

benoitdm-oslandia committed:
fix `QgsOgrLayer::GetExtent3D` by using the real drivername in `QgsOgrProviderUtils::quotedIdentifier` calls

benoitdm-oslandia committed:
fix `QgsOgrProvider` comments

benoitdm-oslandia committed:
fix `QgsOgrProvider::extent()` by reusing `computeExtent3DSlowly` and handling error from `GetExtent`

benoitdm-oslandia committed:
fix maplayer to handle double extents (2D/3D) as for `QgsOgrProvider`

benoitdm-oslandia committed:
fix ogr provider tests to match 2d/3d extent changes

benoitdm-oslandia commented:
fix tests to handle 2d/3d extent changes. 

@rouault There is something strange with the `test_provider_ogr_sqlite.py` test: it seems the `ST_MinZ` and `ST_MaxZ` do not work: they always return 0.0 . Is this expected?

benoitdm-oslandia commented:
@rouault have you an idea why `testExtentSqlite` https://github.com/qgis/QGIS/blob/3b93870b48d5bd72b677b68c52a1d2cc12d33900/tests/src/python/test_provider_ogr_sqlite.py#L475 fails at line 566? 

With `QGIS_DEBUG=3` I saw the function `QgsOgrLayer::GetExtent3D` prints `done with Z! 0.0/0.0` at line https://github.com/qgis/QGIS/blob/3b93870b48d5bd72b677b68c52a1d2cc12d33900/src/core/providers/ogr/qgsogrproviderutils.cpp#L3146

Regards

rouault commented:
> have you an idea why `testExtentSqlite`

yes, I gave you the reason in previous comment ;-)

Change
``ds = ogr.GetDriverByName('SQLite').CreateDataSource(tmpfile)``
to
``ds = ogr.GetDriverByName('SQLite').CreateDataSource(tmpfile, options=["SPATIALITE=YES"])``

But this probably means you should be more careful in the OGR providerl when using ST_MinZ/ST_MaxZ for a SQLite DB. You should like issue a ExecuteSQL("SELECT spatialite_version()") first and see if it returns something or an error

benoitdm-oslandia committed:
fix QgsOgrProvider::extentXX by returning null extent when error occur

benoitdm-oslandia committed:
fix extent 2D/3D for sqlite: handles case with or without spatialite.

rouault commented:
Merging.

benoitdm-oslandia commented:
Thanks @nyalldawson @rouault for the review process!
